src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeReader.java

```diff
@@ -0,0 +1,455 @@
+        this.postingsIndex = new BlockBalancedTreePostingsIndex(postingsFile, treePostingsRoot);
+        leafOrderMapReader = DirectReaders.getReaderForBitsPerValue((byte) DirectWriter.unsignedBitsRequired(maxPointsInLeafNode - 1));
+    }
+
+    public int getBytesPerValue()
+    {
+        return bytesPerValue;
+    }
+
+    public long getPointCount()
+    {
+        return pointCount;
+    }
+
+    @Override
+    public void close()
+    {
+        super.close();
+        FileUtils.closeQuietly(postingsFile);
+    }
+
+    @SuppressWarnings({"resource", "RedundantSuppression"})
+    public PostingList intersect(IntersectVisitor visitor, QueryEventListener.BalancedTreeEventListener listener, QueryContext context)
+    {
+        Relation relation = visitor.compare(minPackedValue, maxPackedValue);
+
+        if (relation == Relation.CELL_OUTSIDE_QUERY)
+        {
+            listener.onIntersectionEarlyExit();
+            return null;
+        }
+
+        listener.onSegmentHit();
+        IndexInput treeInput = IndexFileUtils.instance.openInput(treeIndexFile);
+        IndexInput postingsInput = IndexFileUtils.instance.openInput(postingsFile);
+        IndexInput postingsSummaryInput = IndexFileUtils.instance.openInput(postingsFile);
+        PackedIndexTree index = new PackedIndexTree();
+
+        Intersection intersection = relation == Relation.CELL_INSIDE_QUERY
+                                    ? new Intersection(treeInput, postingsInput, postingsSummaryInput, index, listener, context)
+                                    : new FilteringIntersection(treeInput, postingsInput, postingsSummaryInput, index, visitor, listener, context);
+
+        return intersection.execute();
+    }
+
+    /**
+     * Synchronous intersection of a point or point range with a block balanced tree previously written
+     * with {@link BlockBalancedTreeWriter}.
+     */
+    private class Intersection
+    {
+        private final Stopwatch queryExecutionTimer = Stopwatch.createStarted();
+        final QueryContext context;
+
+        final IndexInput treeInput;
+        final IndexInput postingsInput;
+        final IndexInput postingsSummaryInput;
+        final PackedIndexTree index;
+        final QueryEventListener.BalancedTreeEventListener listener;
+
+        Intersection(IndexInput treeInput, IndexInput postingsInput, IndexInput postingsSummaryInput,
+                     PackedIndexTree index, QueryEventListener.BalancedTreeEventListener listener, QueryContext context)
+        {
+            this.treeInput = treeInput;
+            this.postingsInput = postingsInput;
+            this.postingsSummaryInput = postingsSummaryInput;
+            this.index = index;
+            this.listener = listener;
+            this.context = context;
+        }
+
+        public PostingList execute()
+        {
+            try
+            {
+                PriorityQueue<PeekablePostingList> postingLists = new PriorityQueue<>(100, COMPARATOR);
```

Comment:
This is a hot path, so it seems like spending a little extra effort to size the PQ might be worth it. We do have some metadata about the tree in the `SegmentMetadata`, right? For instance, if we had a tree w/ 60 leaf nodes, and `MINIMUM_POSTINGS_LEAVES` is the default 64, we know that the worst we could do here is size the PQ to 60, rather than 100?

Reply:
nit: It might also be nicer to just make `postingLists` a field on the `Intersection` so we don't have to pass it through so many methods below. This class is basically a "collector"...

Reply:
Agree with both of these. Putting `postingLists` on the `Intersection` makes things a lot clearer and getting the number of leaves isn't any work because we already have that from the index.